File: tally-android/app/src/main/kotlin/app/tally/analytics/TallyAnalytics.kt
package app.tally.analytics

import android.content.Context
import com.posthog.PostHog
import com.posthog.android.PostHogAndroid
import com.posthog.android.PostHogAndroidConfig
import io.sentry.Sentry
import io.sentry.protocol.User

/**
 * Tally Analytics for Android
 *
 * Provides unified analytics tracking with PostHog and error reporting with Sentry.
 * See docs/ANALYTICS.md for the event taxonomy.
 */
object TallyAnalytics {
    @Volatile
    private var isInitialized = false

    /**
     * Initialize analytics (call once from Application.onCreate)
     */
    fun configure(
        context: Context,
        posthogApiKey: String?,
        appVersion: String
    ) {
        if (isInitialized) return

        // Initialize PostHog
        if (!posthogApiKey.isNullOrBlank()) {
            val config = PostHogAndroidConfig(
                apiKey = posthogApiKey,
                host = "https://app.posthog.com",
            ).apply {
                captureScreenViews = false
                captureDeepLinks = false
            }

            PostHogAndroid.setup(context, config)

            PostHog.register("platform", "android")
            PostHog.register("app_version", appVersion)
        }

        // Note: Sentry is auto-initialized by the Gradle plugin
        // Just ensure user is cleared on init
        Sentry.configureScope { scope ->
            scope.user = null
        }

        isInitialized = true
    }

    /**
     * Identify the current user (ID is hashed for privacy)
     */
    fun identify(userId: String, traits: Map<String, Any>? = null) {
        val hashedId = hashUserId(userId)
        PostHog.identify(hashedId, traits ?: emptyMap(), emptyMap())

        val sentryUser = User().apply {
            id = hashedId
        }
        Sentry.setUser(sentryUser)
    }

    /**
     * Clear user identity on sign out
     */
    fun reset() {
        PostHog.reset()
        Sentry.setUser(null)
    }

    /**
     * Track an analytics event
     */
    fun track(event: String, properties: Map<String, Any>? = null) {
        val props = (properties ?: emptyMap()).toMutableMap()
        props["platform"] = "android"

        PostHog.capture(event, null, props)
    }

    // MARK: - Convenience Methods

    fun trackEntryCreated(count: Int, hasNote: Boolean, hasSets: Boolean, hasFeeling: Boolean) {
        track("entry_created", mapOf(
            "count" to count,
            "has_note" to hasNote,
            "has_sets" to hasSets,
            "has_feeling" to hasFeeling
        ))
    }

    fun trackChallengeCreated(timeframeUnit: String, targetNumber: Int, isPublic: Boolean) {
        track("challenge_created", mapOf(
            "timeframe_unit" to timeframeUnit,
            "target_number" to targetNumber,
            "is_public" to isPublic
        ))
    }

    fun trackStreakAchieved(streakDays: Int, challengeId: String) {
        track("streak_achieved", mapOf(
            "streak_days" to streakDays,
            "challenge_id" to challengeId
        ))
    }

    fun trackChallengeViewed(challengeId: String, isOwn: Boolean) {
        track("challenge_viewed", mapOf(
            "challenge_id" to challengeId,
            "is_own" to isOwn
        ))
    }

    // MARK: - Error Tracking

    /**
     * Capture an exception to Sentry
     */
    fun captureException(throwable: Throwable, context: Map<String, Any>? = null) {
        if (context != null) {
            Sentry.configureScope { scope ->
                context.forEach { (key, value) ->
                    scope.setExtra(key, value.toString())
                }
            }
        }
        Sentry.captureException(throwable)
    }

    /**
     * Add a breadcrumb for debugging
     */
    fun addBreadcrumb(message: String, category: String? = null) {
        val breadcrumb = io.sentry.Breadcrumb().apply {
            this.message = message
            this.category = category ?: "app"
        }
        Sentry.addBreadcrumb(breadcrumb)
    }

    // MARK: - Private

    private fun hashUserId(userId: String): String {
        var hash: Long = 5381
        for (char in userId) {
            hash = ((hash shl 5) + hash) + char.code
        }
        return "u_${hash.toULong().toString(36)}"
    }
}

File: tally-ios/TallyCore/Sources/TallyCore/Analytics/TallyAnalytics.swift
import Foundation
import PostHog
import Sentry

/// Tally Analytics for iOS
///
/// Provides unified analytics tracking with PostHog and error reporting with Sentry.
/// See docs/ANALYTICS.md for the event taxonomy.
public final class TallyAnalytics: @unchecked Sendable {
  public static let shared = TallyAnalytics()
  
  private var isInitialized = false
  
  private init() {}
  
  // MARK: - Configuration
  
  public struct Config {
    public let posthogApiKey: String?
    public let sentryDsn: String?
    public let environment: String
    public let appVersion: String
    
    public init(
      posthogApiKey: String? = nil,
      sentryDsn: String? = nil,
      environment: String = "production",
      appVersion: String
    ) {
      self.posthogApiKey = posthogApiKey
      self.sentryDsn = sentryDsn
      self.environment = environment
      self.appVersion = appVersion
    }
  }
  
  // MARK: - Initialization
  
  /// Initialize analytics and error tracking
  public func configure(_ config: Config) {
    guard !isInitialized else { return }
    
    // Initialize PostHog
    if let posthogKey = config.posthogApiKey, !posthogKey.isEmpty {
      let posthogConfig = PostHogConfig(apiKey: posthogKey)
      posthogConfig.captureScreenViews = false
      posthogConfig.captureApplicationLifecycleEvents = true
      PostHogSDK.shared.setup(posthogConfig)
      PostHogSDK.shared.register(["platform": "ios", "app_version": config.appVersion])
    }
    
    // Initialize Sentry
    if let sentryDsn = config.sentryDsn, !sentryDsn.isEmpty {
      SentrySDK.start { options in
        options.dsn = sentryDsn
        options.environment = config.environment
        options.releaseName = config.appVersion
        options.tracesSampleRate = 0.1
        options.enableAutoSessionTracking = true
        
        // Privacy: don't capture PII
        options.beforeSend = { event in
          event.user?.email = nil
          event.user?.username = nil
          return event
        }
      }
    }
    
    isInitialized = true
  }
  
  // MARK: - User Identification
  
  /// Identify the current user (ID is hashed for privacy)
  public func identify(userId: String, traits: [String: Any]? = nil) {
    let hashedId = hashUserId(userId)
    PostHogSDK.shared.identify(hashedId, userProperties: traits)
    
    let sentryUser = Sentry.User()
    sentryUser.userId = hashedId
    SentrySDK.setUser(sentryUser)
  }
  
  /// Clear user identity on sign out
  public func reset() {
    PostHogSDK.shared.reset()
    SentrySDK.setUser(nil)
  }
  
  // MARK: - Event Tracking
  
  /// Track an analytics event
  public func track(_ event: String, properties: [String: Any]? = nil) {
    var props = properties ?? [:]
    props["platform"] = "ios"
    
    PostHogSDK.shared.capture(event, properties: props)
  }
  
  // MARK: - Convenience Methods
  
  public func trackEntryCreated(count: Int, hasNote: Bool, hasSets: Bool, hasFeeling: Bool) {
    track("entry_created", properties: [
      "count": count,
      "has_note": hasNote,
      "has_sets": hasSets,
      "has_feeling": hasFeeling
    ])
  }
  
  public func trackChallengeCreated(timeframeUnit: String, targetNumber: Int, isPublic: Bool) {
    track("challenge_created", properties: [
      "timeframe_unit": timeframeUnit,
      "target_number": targetNumber,
      "is_public": isPublic
    ])
  }
  
  public func trackStreakAchieved(streakDays: Int, challengeId: String) {
    track("streak_achieved", properties: [
      "streak_days": streakDays,
      "challenge_id": challengeId
    ])
  }
  
  public func trackChallengeViewed(challengeId: String, isOwn: Bool) {
    track("challenge_viewed", properties: [
      "challenge_id": challengeId,
      "is_own": isOwn
    ])
  }
  
  // MARK: - Error Tracking
  
  /// Capture an error to Sentry
  public func captureError(_ error: Error, context: [String: Any]? = nil) {
    if let context = context {
      SentrySDK.configureScope { scope in
        for (key, value) in context {
          scope.setExtra(value: value, key: key)
        }
      }
    }
    SentrySDK.capture(error: error)
  }
  
  /// Add a breadcrumb for debugging
  public func addBreadcrumb(message: String, category: String? = nil) {
    let crumb = Breadcrumb()
    crumb.message = message
    if let category { crumb.category = category }
    crumb.level = .info
    SentrySDK.addBreadcrumb(crumb)
  }
  
  // MARK: - Private
  
  private func hashUserId(_ userId: String) -> String {
    var hash: UInt64 = 5381
    for char in userId.utf8 {
      hash = ((hash << 5) &+ hash) &+ UInt64(char)
    }
    return "u_\(String(hash, radix: 36))"
  }
}
